doc/designs/lms.md

```diff
@@ -0,0 +1,194 @@
+- Dont support streaming via existing interface:
+  EVP_DigestVerifyInit_ex + EVP_DigestVerifyUpdate + EVP_DigestVerifyFinal.
+  It could be done, but requires buffering of the message during updates.
+- EVP_VerifyInit etc should not be supported.
+- Add streaming via a new API that is similar to the existing DigestVerify API's,
+  but moves the signature |sig| from the final to the init. i.e.
+
+```c
+__owur int EVP_DigestVerifyHBSInit(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
+                                   const char *mdname,
+                                   OSSL_LIB_CTX *libctx,
+                                   const char *props, EVP_PKEY *pkey,
+                                   const OSSL_PARAM params[],
+                                   const unsigned char *sig,
+                                   size_t siglen);
+   
+/* Alias: since the function signature is identical */
+#define EVP_DigestVerifyHBSUpdate(a,b,c) EVP_DigestVerifyUpdate(a,b,c)
+
+__owur int EVP_DigestVerifyHBSFinal(EVP_MD_CTX *ctx);
+
+```
+
+Hash Function Selection
+-----------------------
+
+- During the DigestVerifyInit()the passed in digest name is not required
+since the LMS/HSS public key and HSS signature contain the digest to use.
+If specified, the digest name must match the one used by the HSS public key and
+signature, otherwise an error will occur.
+- Section 4 of [2] states that the same hash function MUST be used everywhere
+for OTS and LMS at all HSS levels (this means the terms 'm' and 'n' are the
+same)). ([1] only states that the hashes SHOULD be the same).
+This requires error checking when the HSS signature data is decoded.
+- Hash Functions for SHAKE and truncated SHA256 are not specified in [1],
+but are part of [2] so they should be supported.
+For truncated SHAKE256/192 support this will require the OSSL_DIGEST_PARAM_XOFLEN
+parameter to be set.
+
+Threading
+---------
+
+For HSS, multi-level trees are used for verification,
+and there is a verify process involving many hashes for each of the (1..8)
+levels, with only the final level using the input message (and the message is
+possibly streamed).
+All the other levels are one shot verify operations that must all return true in
+order to pass verification. These one-shot operations may be done in parallel
+using the new internal thread API calls (introduced for ARGON).
+When streaming only the final part of the validation involves the message,
+and this is done using a init, update and final in the main thread.
+After this the other threads must finish before verification can end.
+If threads are not supported, then a single threaded call is used instead.
+By default the thread count will be one.
+A parameter to set the threads can be done via a OSSL_PARAM.
+For one shot verification threading can still be used for the top levels.
+
+Provider Support
+----------------
+
+dupctx should not be supported by the provider since this would not apply to LMS.
```

Comment:
This section is a bit too sparse.  I'd like to know what you see us adding in the libcrypto\<->provider interface (i.e. in core_dispatch.h)

Reply:
Added..